Performance and api improvements for messaging

davidchappelle opened the pull request:
This commit converts all messages to use shallow copy semantics
for unmarshalling msgpack objects rather than deep copying each
object.

davidchappelle committed:
Performance and api improvements for messaging

This commit converts all messages to use shallow copy semantics
for unmarshalling msgpack objects rather than deep copying each
object.

jpetso commented on src/bonefish/serialization/json_serializer.cpp:
```diff
@@ -167,11 +167,11 @@ struct wamp_bin_string_conversion
 
     imemstream bufferstream(buffer, length);
-    serialization::msgpack_from_json_handler<wamp_bin_string_conversion> handler(item, zone);
+    serialization::msgpack_from_json_handler<wamp_bin_string_conversion> handler(item, *(zone.get()));
```
`*zone` is a more concise way to ask for a reference to the object in a smart pointer.

jpetso commented on src/bonefish/messages/wamp_message.hpp:
```diff
@@ -28,22 +29,47 @@ class wamp_message
     wamp_message& operator=(wamp_message const&) = delete;
+    wamp_message& operator=(wamp_message&&) = delete;
+
+    std::unique_ptr<msgpack::zone>&& release_zone();
```
I think it would be better to return a new unique_ptr object, not a reference to the internal one. When I call `release_zone()`, I expect the zone to be gone regardless of whether I decide to move it into another pointer object or not. It shouldn't make a noticeable (or any, with compiler optimizations?) difference to performance.

davidchappelle replied:
I like the idioms of release/acquire. Do you have a better suggestion for naming?

jpetso replied:
Nope, release/acquire is great. My suggestion was just to make release actually release the pointer rather than just providing a movable reference to it. For clarification, I'm thinking of a function signature like this:
`std::unique_ptr<msgpack::zone> release_zone();`
or in case you decide to use the zone's built-in move assignment, maybe this:
`msgpack::zone&& move_zone();`

davidchappelle replied:
Ok cool. I originally wanted to avoid unique_ptr and just use zone's built in move semantics however, msgpack::unpacker stores its zone as a unique_ptr so unfortunately we are stuck with unique_ptr for the time being.

jpetso replied:
That by itself is not really much of a hindrance, as you could just do `std::move(*unpacker.zone())` instead of moving its pointer :P but I'll leave it up to you which one to prefer in the end.

davidchappelle replied:
I like it. Let me see what I can do as unique_ptr was a painful thing to carry around.

jpetso commented on src/bonefish/messages/wamp_message.hpp:
```diff
@@ -28,22 +29,47 @@ class wamp_message
+    wamp_message& operator=(wamp_message&&) = delete;
+
+    std::unique_ptr<msgpack::zone>&& release_zone();
+    const std::unique_ptr<msgpack::zone>& get_zone() const;
```
If you're not expecting the caller to modify or move the unique_ptr then you might as well return the actual zone pointer directly. I think C++17 will have a type to use for non-owned pointers, but for now a plain pointer is probably the best we have for this use case.

jpetso replied:
Actually, maybe you even want `msgpack::zone& get_zone() const;` and save the deferencing star that you're currently adding everywhere. Of course this would crash, but so would the current code if the zone pointer is null.

davidchappelle replied:
So if I ditch unique_ptr would this actually then become:

msgpack::zone release_zone() { return std::move(m_zone); }
msgpack::zone& get_zone() { return m_zone; }

jpetso replied:
That would work, yes.

jpetso commented:
This seems like a worthwhile change to me. I noticed that you're not actually using `wamp_message::release_zone()` anywhere, which makes me wonder whether it would be slightly nicer to use `msgpack::zone`'s built-in move operator instead of allocating the zone with an extra layer of pointer indirection. Either way, please have a look at my line comments. Thanks!

davidchappelle commented:
Thanks Jacob. I will be using release_zone when I submit a pull request for
#20 as well as #21.

On Thu, Oct 8, 2015 at 2:58 PM, Jakob Petsovits notifications@github.com
wrote:

> This seems like a worthwhile change to me. I noticed that you're not
> actually using wamp_message::release_zone() anywhere, which makes me
> wonder whether it would be slightly nicer to use msgpack::zone's built-in
> move operator instead of allocating the zone with an extra layer of pointer
> indirection. Either way, please have a look at my line comments. Thanks!
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/tplgy/bonefish/pull/19#issuecomment-146654554.

davidchappelle committed:
Add support for native components

This commit adds the necessary infrastructure required to allow
for components to connect natively to bonefish and join a session
with a wamp router. This should allow for some very high performance
messaging as native support was added with a zero copy philosophy. A
message that is sent to bonefish over a native transport is only ever
shallow copied according to the msgpack::object and msgpack::zone
semantics. A native transport implementation is still in progress using
AutobahnCpp and should be too far behind this commit. It will be used to
finish testing. I would consider native support to be experimental at this
stage and it is not yet enabled in the bonefish daemon.

Fixes: #21

jpetso commented:
Commit message: "A native transport implementation is still in progress using AutobahnCpp and should be too far behind this commit." -> "shouldn't be"?

jpetso commented:
Seems sensible in general! I left a bunch of comments, most of them only related to style/documentation. Please have a look and report back!
